In [1]:
import numpy as np
import pandas as pd
from sklearn.datasets import load_iris

# Load dataset Iris dari scikit-learn
iris = load_iris()

pada cell kode berikut digunakan untuk memanggil library yang diperlukan dalam proses klasifikasi menggunakan algoritma Naive Bayes. Numpy digunakan untuk melakukan operasi pada data numerik, sedangkan Pandas digunakan untuk mengelola dataset dalam bentuk DataFrame. Selanjutnya load_iris digunakan untuk mengambil dataset iris yang sudah tersedia di Scikit-learn. Dataset Irisi digunakan karena memiliki empat fitur numerik yaitu panjang sepal, lebar sepal, panjang petal dan lebar petal.

In [2]:
# Konversi data menjadi DataFrame Pandas
df = pd.DataFrame(iris.data, columns=iris.feature_names)
df["Species"] = iris. target_names [iris.target]

# Definisikan X (fitur) dan y (target/Label)
X = df.drop(columns=["Species"])
y = df["Species"]

print(X)

     sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                  5.1               3.5                1.4               0.2
1                  4.9               3.0                1.4               0.2
2                  4.7               3.2                1.3               0.2
3                  4.6               3.1                1.5               0.2
4                  5.0               3.6                1.4               0.2
..                 ...               ...                ...               ...
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

[150 rows x 4 columns]


Pada cell ini dataset Iris memiliki 150 data dengan 4 fitur, yaitu sepal length (cm), sepal width (cm), petal length (cm), dan petal width (cm). Sementara itu, variabel Species digunakan sebagai target atau kelas yang terdiri dari tiga jenis bunga Iris, yaitu setosa, versicolor, dan virginica. Dengan demikian, X berisi empat fitur yang digunakan sebagai dasar prediksi, sedangkan y berisi jenis species yang akan diprediksi.

In [3]:
y

0         setosa
1         setosa
2         setosa
3         setosa
4         setosa
         ...    
145    virginica
146    virginica
147    virginica
148    virginica
149    virginica
Name: Species, Length: 150, dtype: str

Pada cell berikut digunakan untuk mengubah nama species yang awalnya berupa teks menjadi nilai numerik agar dapat diproses oleh model machine learning. Hasil encoding menunjukkan bahwa setosa diubah menjadi nilai 0, versicolor menjadi 1, dan virginica menjadi 2. Berdasarkan output praktikum, 50 data pertama memiliki label 0, 50 data berikutnya memiliki label 1, dan 50 data terakhir memiliki label 2

In [4]:
from sklearn.preprocessing import LabelEncoder
lab = LabelEncoder()
y = lab.fit_transform(y)
y

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2])

pada cell tersebut digunakan untuk mengubah nama species yang awalnya berupa teks menjadi nilai numerik agar dapat diproses oleh model machine learning. Hasil encoding menunjukkan bahwa setosa diubah menjadi nilai 0, versicolor menjadi 1, dan virginica menjadi 2. Berdasarkan output praktikum, 50 data pertama memiliki label 0, 50 data berikutnya memiliki label 1, dan 50 data terakhir memiliki label 2

In [5]:
from sklearn.model_selection import train_test_split

X_train, X_test = train_test_split(X, test_size=0.2, random_state=0)
y_train, y_test = train_test_split(y, test_size=0.2, random_state=0)

 pada cell berikut pembagian data dilakukan dengan menggunakan test_size=0.2, sehingga 20% data digunakan sebagai data testing dan 80% digunakan sebagai data training. Dengan jumlah dataset sebanyak 150 data, secara umum terdapat 120 data untuk training dan 30 data untuk testing. random_state=0 digunakan agar pembagian data tetap sama ketika kode dijalankan kembali. Tahapan pembagian data ini diperlukan agar model dapat dilatih menggunakan data training dan kemudian diuji menggunakan data yang tidak digunakan dalam proses pelatihan.

In [6]:
from sklearn.naive_bayes import GaussianNB

model = GaussianNB()
NB_model = model. fit(X_train, y_train)

pada cell kode tersebut membuat model Gaussian Naïve Bayes menggunakan GaussianNB(). Selanjutnya, fit() digunakan untuk melatih model berdasarkan X_train sebagai data fitur dan y_train sebagai label atau target. Setelah proses training selesai, model disimpan dalam variabel NB_model sehingga dapat digunakan untuk melakukan prediksi terhadap data testing maupun data baru.

In [8]:
result = NB_model.predict(X_test)
result

array([2, 1, 0, 2, 0, 2, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 2, 1,
       0, 0, 2, 0, 0, 1, 1, 0])

pada cell berikut hasil prediksi berupa array angka yang menunjukkan kelas dari masing-masing data testing. Nilai 0 menunjukkan bahwa data diprediksi sebagai setosa, nilai 1 menunjukkan versicolor, sedangkan nilai 2 menunjukkan virginica. Pada praktikum, hasil prediksi terhadap 30 data testing ditampilkan dalam bentuk array yang terdiri dari nilai 0, 1, dan 2

In [9]:
from sklearn.metrics import accuracy_score
print('Accuracy of data testing: ', accuracy_score(y_test, result)*100)

Accuracy of data testing:  96.66666666666667


pada cell model Gaussian Naïve Bayes memperoleh accuracy sebesar 96.66666666666667%, atau jika dibulatkan menjadi sekitar 96,67%. Hasil tersebut menunjukkan bahwa model mampu melakukan klasifikasi dengan sangat baik terhadap data testing. Dari 30 data testing, sekitar 29 data berhasil diklasifikasikan dengan benar dan hanya sekitar satu data yang mengalami kesalahan klasifikasi.

In [12]:
#prediksi inputan data baru
#input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4,0,2]]), columns=iris.feature_names)
input_data = [[5.1, 3.5, 1.4, 0.2]]
result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)

Hasil prediksi untuk input baru adalah: ['setosa']


c:\Users\sthab\AppData\Local\Python\pythoncore-3.14-64\Lib\site-packages\sklearn\utils\validation.py:2830: UserWarning: X does not have valid feature names, but GaussianNB was fitted with feature names
  warnings.warn(


pada cell kode tersebut membuat satu data baru dengan empat nilai fitur, yaitu sepal length = 7.3 cm, sepal width = 3.8 cm, petal length = 4.6 cm, dan petal width = 2.3 cm. Data tersebut kemudian diberi nama kolom yang sama dengan dataset Iris sehingga format input sesuai dengan format data yang digunakan ketika model Gaussian Naïve Bayes dilatih. Praktikum juga menjelaskan bahwa penggunaan DataFrame dengan nama kolom dapat menghindari warning yang muncul ketika model dilatih menggunakan fitur yang memiliki nama kolom tetapi prediksi dilakukan menggunakan array tanpa nama kolom

In [13]:
input_data = pd.DataFrame(np.array([[5.1, 3.5, 1.4, 0.2]]), columns=iris.feature_names)

result = NB_model.predict(input_data)
prediction_label = lab.inverse_transform(result)
print('Hasil prediksi untuk input baru adalah:', prediction_label)


Hasil prediksi untuk input baru adalah: ['setosa']


pada cell berikut kode NB_model.predict(input_data) digunakan untuk menentukan kelas berdasarkan empat fitur yang diberikan. Karena hasil prediksi masih berupa angka hasil encoding, yaitu 0, 1, atau 2, maka digunakan inverse_transform() untuk mengembalikan hasil tersebut menjadi nama species aslinya. Dengan demikian, hasil yang ditampilkan tidak lagi berupa angka tetapi berupa nama species seperti setosa, versicolor, atau virginica.

In [14]:
#TUGAS PRAKTIKUM
#42324042_Tamara Sinaga

import numpy as np
import pandas as pd

from sklearn.datasets import load_iris
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import accuracy_score

# Load dataset Iris
iris = load_iris()

# Membuat DataFrame
df = pd.DataFrame(iris.data, columns=iris. feature_names)
df["Species"] = iris.target_names[iris. target]

# Menentukan fitur don target
X = df.drop(columns=["Species"])
y = df["Species"]

# Encoding target
lab = LabelEncoder()
y = lab. fit_transform(y)

# Membagi data training dan testing
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=0
)

# Membuat dan melatih model Gaussian Naive Bayes
model = GaussianNB()
NB_model = model. fit(X_train, y_train)

# Prediksi data testing
result = NB_model.predict(X_test)

# Menghitung accuracy
accuracy = accuracy_score(y_test, result) * 100
print("Accuracy of data testing:", accuracy)

# Data baru sesuai tugas
input_data = pd.DataFrame(
    np.array([[7.3, 3.8, 4.6, 2.3]]),
    columns=iris. feature_names
    )

# Prediksi data baru
result = NB_model.predict(input_data)

# Mengubah angka prediksi menjadi nama species
prediction_label = lab.inverse_transform(result)

print("Hasil prediksi untuk input baru adalah:", prediction_label)

Accuracy of data testing: 96.66666666666667
Hasil prediksi untuk input baru adalah: ['virginica']


pada cell berikut algoritma Gaussian Naïve Bayes dapat digunakan untuk melakukan klasifikasi pada dataset Iris karena seluruh fitur yang digunakan merupakan data numerik kontinu. Model yang dilatih memperoleh akurasi pengujian sekitar 96,67%, sehingga dapat dikatakan memiliki kemampuan klasifikasi yang baik pada data Iris. Untuk tugas yang diberikan, yaitu melakukan prediksi terhadap data baru dengan nilai [7.3, 3.8, 4.6, 2.3], model menghasilkan prediksi virginica. Dengan demikian, jawaban dari tugas praktikum adalah bahwa data baru tersebut masuk ke dalam species Iris virginica